Monika After Story/game/zz_apikeys.rpy

```diff
@@ -0,0 +1,556 @@
+# NOTE: we do not store keys in persistent. 
+#   this is for safety reasons since persistents may be shared online for
+#   troubleshooting purposes.
+
+
+screen mas_update_cert(screen_data):
+
+    # cert update screen
+
+    modal True
+    zorder 200
+    style_prefix "hkb"
+
+    add mas_getTimeFile("gui/overlay/confirm.png")
+
+    timer 1.0 repeat True action Function(renpy.restart_interaction)
+
+    python:
+        # cert checks
+        screen_data.loop()
```

Comment:
the `loop` function doesn't loop, it has handling for if it gets run in a loop, which is what this is expecting the screen to do. Regarding side effects, I think that recommendation is geared toward vn creators, especially when the docs say "Screens must not cause side effects **that are visible from outside the screen**". In this case, the cert update's side effect is downloading the cert and updating the lib, which is not a visible thing, and the loop code is using the async wrapper which has safeguards against running at the wrong times.

`on show` and a button press both don't work for this since the "loop" function would need to be changed to be an actual loop which defeats the purpose of using a screen - the screen is a free "waiting loop" that renpy manages for us so we don't cause the game to hang while doing I/O/internet connecting.

Reply:
> I think that recommendation is geared toward vn creators

It's general advice on how renpy works. It looks odd designing a function to work from inside a screen when renpy doesn't support that and warns us about it. Feels like we're overcomplicating things for no gain. I understand what you did in that function and why - to make it work with a screen, but that's completely against renpy's logic, and when people who don't know that much would look at it, they would be confused or think that's okay to do.

This isn't the supported way of doing things, if in the future renpy changes how it handles that, we'd have no say in that and we'd have to update this piece of code.

It's fine to use a screen for a "waiting loop," we just need to start the function from some event (like mouse click).

Reply:
Are you concerned with the `loop` function in the base class? because for this instance, the start of the thread happens when the Update Cert button is clicked, which runs `mas_api_keys.screen_update_cert` - this function also shows the screen via a renpy.show_screen call (so not predicted under usual renpy rules). 

I could remove the base class `loop` if you are concerned with that being used incorrectly, but it needs to stay for this particular screen.

Reply:
My concern is the `loop` function will be called at random times, sometimes multiple times in a row.
I think after reading the code a few more times, I like it more, although, I still think it's a bit hacky with how we use screen updates. We force all the screens to redraw to randomly call the `loop` function to change screen state. I do like how you encapsulated screen data, but I think it should be put in an empty displayable which would redraw itself (or even better launch pseudo-event) and update self and the screen elements. But even w/o that...

What you have now could be cleaned up more with just
```renpy
timer 1.0 repeat True action Function(screen_data.loop)
```
Does exactly the same, but more precisely and it's more clear what your goal is - calling loop every second. You might or might not need to still restart interaction, which can be done either in the `loop` func or by running 2 `Function` actions via the timer.

Reply:
agreed with the function option, looks cleaner, done.